src/main/kotlin/net/maizegenetics/phgv2/pathing/DiploidPathFinding.kt

```diff
@@ -0,0 +1,325 @@
+import com.github.ajalt.clikt.parameters.types.int
+import htsjdk.variant.variantcontext.VariantContext
+import htsjdk.variant.vcf.VCFAltHeaderLine
+import htsjdk.variant.vcf.VCFHeaderVersion
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.channels.ReceiveChannel
+import kotlinx.coroutines.channels.SendChannel
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.runBlocking
+import net.maizegenetics.phgv2.api.HaplotypeGraph
+import net.maizegenetics.phgv2.utils.*
+import org.apache.logging.log4j.LogManager
+import java.io.File
+
+/**
+ * Finds the most likely pair of paths through a HaplotypeGraph based on read mapping data.
+ * Input: a keyfile with taxa names and read mapping file paths and a haplotype graph. Read mappings must be in the form of
+ * haplotype list -> count of reads mapping to the haplotypes. The haplotype graph must be the one used to map reads or
+ * a subset of that restricted to specific taxa and/or ReferenceRanges.
+ *
+ * Steps:
+ * 1. Read the keyfile
+ * 2. Build the HaplotypeGraph
+ * 3. For each entry in the keyfile (multithreaded)
+ *      a. Get the read mapping data
+ *      b. Use Viterbi algorithm to find the best path
+ *      c. Store the path (write hvcf)
+ *
+ * The samples in the keyfile are processed in parallel. Each thread (co-routine) processes a sample. Within sample
+ * processing is not multithreaded.
+ */
+class DiploidPathFinding: CliktCommand(help = "Impute best diploid path using read mappings.") {
+    val pathKeyfile by option(help = "tab-delimited file with first two columns: SampleName, ReadMappingFiles. ReadMappingFiles " +
+            "must be the full path to a read mapping file or a comma separated list of file paths.")
+        .required()
+        .validate { require(File(it).exists()) {"$it is not a valid file"} }
+
+    val hvcfDir by option(help = "The directory containing the hvcf files used to build a HaplotypeGraph for path finding.")
+        .required()
+        .validate { require(File(it).isDirectory) {"$it is not a valid directory."} }
+
+    val referenceGenome by option(help = "path to reference genome (fasta or fastq")
+        .required()
+        .validate { require(File(it).exists()) {"$it is not a valid file"} }
+
+    val outputDir by option(help = "The directory where the output hvcfs will be written. The output file names will be <sampleName>.h.vcf.")
+        .required()
+        .validate { require(File(it).isDirectory) {"$it is not a valid directory."} }
+
+    val probCorrect by option(help = "The probability that a mapped read was mapped correctly")
+        .double()
+        .default(0.99)
+        .validate { require(it in 0.5..1.0) {"prob-correct must be between 0.5 and 1.0"} }
+
+    val probSameGamete by option(help = "The probability of transitioning to the same gamete in the next reference range")
+        .double()
+        .default(0.99)
```

Comment:
If user provides a number > 1.0, does it default to 1.0?  OR is this an error?

Reply:
It should throw an error due to the require statement.  I believe this is tested in the unit tests.